doc/user-manual/language/core-language.lagda.rst

```diff
@@ -8,21 +8,177 @@
+data structure that we will call Concrete Syntax, while checking for correct
+syntax.
+
+The parser is generated by Happy_.
+
+Example: when a name is a sequence of parts, the lexer just sees it as a
+string, the parser does the translation in this step.
+
+
+Concrete Syntax
+---------------
+
+The concrete syntax is a raw representation of the program text without any
+desugaring at all.  This is what the parser produces. The idea is that if we
+figure out how to keep the concrete syntax around, it can be printed exactly
+as the user wrote it.
+
+Nice Concrete Syntax
+--------------------
+
+The translation from "Agda.Syntax.Concrete" to "Agda.Syntax.Abstract" involves
+scope analysis, figuring out infix operator precedences and tidying up
+definitions.
+
+Abstract Syntax
+---------------
+
+The abstract syntax ``Agda.Syntax.Abstract`` is the result after desugaring
+and scope analysis of the concrete syntax. The type checker works on abstract
+syntax, producing internal syntax.
+
+Internal Syntax
+---------------
+
+This is the final stage of syntax before being handed off to one of the
+backends. Terms are well-scoped and well-typed.
+
+The internal syntax ``Agda.Syntax.Internal`` uses the following haskell
+datatype to represent the grammar of a ``Term`` presented above.
 
 .. code-block:: haskell
 
-  data Term = Var Int Elims
-            | Def QName Elims               -- ^ @f es@, possibly a delta/iota-redex
-            | Con ConHead Args              -- ^ @c vs@
-            | Lam ArgInfo (Abs Term)        -- ^ Terms are beta normal. Relevance is ignored
-            | Lit Literal
-            | Pi (Dom Type) (Abs Type)      -- ^ dependent or non-dependent function space
-            | Sort Sort
-            | Level Level
-            | MetaV MetaId Elims
-            | DontCare Term
-              -- ^ Irrelevant stuff in relevant position, but created
-              --   in an irrelevant context.
+  data Term = Var {-# UNPACK #-} !Int Elims -- ^ @x es@ neutral
+          | Lam ArgInfo (Abs Term)        -- ^ Terms are beta normal. Relevance is ignored
+          | Lit Literal
+          | Def QName Elims               -- ^ @f es@, possibly a delta/iota-redex
+          | Con ConHead ConInfo Elims
+          -- ^ @c es@ or @record { fs = es }@
+          --   @es@ allows only Apply and IApply eliminations,
+          --   and IApply only for data constructors.
+          | Pi (Dom Type) (Abs Type)      -- ^ dependent or non-dependent function space
+          | Sort Sort
+          | Level Level
+          | MetaV {-# UNPACK #-} !MetaId Elims
+
+Treeless Syntax
+---------------
+
+The treeless syntax is intended to be used as input for the
+:ref:`compiler backends <compiler-backends>`. It is more low-level than the
+internal syntax and is not used for type checking. Some of the features of
+the treeless syntax are:
+
+* case expressions instead of case trees
+* no instantiated datatypes / constructors
+
```

Comment:
An interesting feature of treeless is that it is in so-called [`A normal form`](https://en.wikipedia.org/wiki/A-normal_form).

That means that all the case expressions are targeting a *single*
variable and all alternatives may only peel off *one* constructor.
This make it a lot simpler to handle in the backends than a language
where you may case on arbitrary expressions & use deep patterns.